File: libs-node/erouterClient.js
"use strict";
var request = require('request');
var uuid = require('node-uuid');

var service_url;
var publish_retry_count = 10;

var poll = function(topic_name, subscription_name, next) {
    request.get(service_url + "/topic/" + topic_name + "/" + subscription_name + "/poll", function(err, response, body) {
        if (err) {
            setTimeout(function() {
                poll(topic_name, subscription_name, next);
            }, 100);
            next(err);
        } else {
            try {
                var json_blob = JSON.parse(body);
                if (json_blob.message) {
                    var msg = JSON.parse(json_blob.message);
                    request.del(service_url + "/topic/" + topic_name + "/" + subscription_name + "/delete/" + msg['id'], function(err) {
                        if (err) {
                            next(err);
                        } else {
                            next(null, msg['msg']);
                        }
                        poll(topic_name, subscription_name, next);
                    });
                } else {
                    setTimeout(function() {
                        poll(topic_name, subscription_name, next);
                    }, 1000);
                }
            } catch (err) {
                setTimeout(function() {
                    poll(topic_name, subscription_name, next);
                }, 100);
                next(err);
            }
        }
    });
};

var subscribe = function(topic_name, subscription_name, next) {
    subscription_name  += '_' + uuid.v4();
    request.post(service_url + "/topic/" + topic_name + "/" + subscription_name + "/subscribe", function(err) {
        if (err) {
            next(err);
        } else {
            poll(topic_name, subscription_name, next);
        }
    });
};

var publish = function(topic_name, message, next, retries) {
    retries = typeof retries !== 'undefined' ? retries : 0;
    request.put({
        headers: {
            'content-type': 'application/json'
        },
        url: service_url + '/topic/' + topic_name + '/publish',
        body: message
    }, function(err) {
        if (err) {
            if (retries < publish_retry_count) {
                retries++;
                setTimeout(function() {
                    publish(topic_name, message, next, retries);
                }, 100);
            } else {
                next(err);
            }
        } else {
            next();
        }
    });
};

exports.subscribe = subscribe;
exports.publish = publish;
exports.set_service_url = function(url) {
    service_url = url;
};

File: lwutils/erouter_client.py
import json
from threading import Thread
import requests
import time
from time import sleep
from uuid import uuid4


class ERouterClient:

    def __init__(self, erouter_url):
        self.url = erouter_url
        self.publish_retry_count = 10

    def _poll(self, topic_name, subscription_name, callback):
        while True:
            try:
                response = requests.get("%s/topic/%s/%s/poll" % (self.url, topic_name, subscription_name))
                if response.status_code == 200:
                    message = json.loads(response.text).get("message")
                    if message:
                        message_dict = json.loads(message)
                        del_response = requests.delete(
                            "%s/topic/%s/%s/delete/%s" % (self.url, topic_name, subscription_name, message_dict['id']))
                        if del_response.status_code == 200:
                            callback(message_dict['msg'])
                        continue
                    else:
                        time.sleep(2)
            except Exception as e:
                print 'Error during callback for event %s and subscription %s. Error: %s' % (topic_name, subscription_name, e)

    def subscribe(self, topic_name, subscription_name, callback=None, worker_mode=False):
        if not worker_mode:
            subscription_name += "_" + str(uuid4())  # make subscription name unique for scaling
        requests.post("%s/topic/%s/%s/subscribe" % (self.url, topic_name, subscription_name))
        if callback:
            t = Thread(target=self._poll, args=(topic_name, subscription_name, callback,))
            t.daemon = True
            t.start()

    def publish(self, topic_name, message, retry=0):
        try:
            requests.put("%s/topic/%s/publish" % (self.url, topic_name), data=message)
        except Exception as e:
            if retry < self.publish_retry_count:
                sleep(0.1)
                self.publish(topic_name, message, retry=retry + 1)
            else:
                raise e

    def create(self, topic_name):
        requests.put("%s/topic/%s/create" % (self.url, topic_name))
